app/scripts/metamask-controller.js

```diff
@@ -5071,56 +5034,137 @@ export default class MetamaskController extends EventEmitter {
+        return 0;
+      } else if (firstAccount.metadata.lastSelected === undefined) {
+        return 1;
+      } else if (secondAccount.metadata.lastSelected === undefined) {
+        return -1;
+      }
+
+      return (
+        secondAccount.metadata.lastSelected - firstAccount.metadata.lastSelected
+      );
+    });
+  }
+
   /**
    * Gets the permitted accounts for the specified origin. Returns an empty
    * array if no accounts are permitted.
    *
    * @param {string} origin - The origin whose exposed accounts to retrieve.
-   * @param {boolean} [suppressUnauthorizedError] - Suppresses the unauthorized error.
    * @returns {Promise<string[]>} The origin's permitted accounts, or an empty
    * array.
    */
-  async getPermittedAccounts(
-    origin,
-    { suppressUnauthorizedError = true } = {},
-  ) {
+  getPermittedAccounts(origin) {
+    let caveat;
     try {
-      return await this.permissionController.executeRestrictedMethod(
+      caveat = this.permissionController.getCaveat(
         origin,
-        RestrictedMethods.eth_accounts,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
       );
-    } catch (error) {
-      if (
-        suppressUnauthorizedError &&
-        error.code === rpcErrorCodes.provider.unauthorized
-      ) {
-        return [];
-      }
-      throw error;
+    } catch (err) {
+      // noop
     }
+    if (!caveat) {
+      return [];
+    }
+
+    return getEthAccounts(caveat.value);
+  }
+
+  async getPermittedAccountsSorted(origin) {
+    const permittedAccounts = this.getPermittedAccounts(origin);
+    const allEvmAccounts = await this.getAllEvmAccountsSorted();
+    return allEvmAccounts.filter((account) =>
+      permittedAccounts.includes(account),
+    );
   }
 
   /**
    * Stops exposing the specified chain ID to all third parties.
-   * Exposed chain IDs are stored in caveats of the `endowment:permitted-chains`
-   * permission. This method uses `PermissionController.updatePermissionsByCaveat`
-   * to remove the specified chain ID from every `endowment:permitted-chains`
-   * permission. If a permission only included this chain ID, the permission is
-   * revoked entirely.
    *
    * @param {string} targetChainId - The chain ID to stop exposing
    * to third parties.
    */
   removeAllChainIdPermissions(targetChainId) {
     this.permissionController.updatePermissionsByCaveat(
-      CaveatTypes.restrictNetworkSwitching,
-      (existingChainIds) =>
-        CaveatMutatorFactories[
-          CaveatTypes.restrictNetworkSwitching
-        ].removeChainId(targetChainId, existingChainIds),
+      Caip25CaveatType,
+      (existingScopes) =>
+        Caip25CaveatMutators[Caip25CaveatType].removeScope(
+          existingScopes,
+          toCaipChainId('eip155', parseInt(targetChainId, 16).toString()),
+        ),
     );
   }
 
+  removeNetwork(chainId) {
+    this.removeAllChainIdPermissions(chainId);
+
+    this.networkController.removeNetwork(chainId);
+  }
+
```

Comment:
When there are two networks that serve the same chainId, removing only one of the networks will still cause that chain to be revoked from all permissions. This is wrong of course.

This is a bug that needs to be resolved in `main` (where removing any network doesn't cause chain permissions to be updated)

Reply:
Bug ticket [here](https://app.zenhub.com/workspaces/wallet-api-platform-63bee08a4e3b9d001108416e/issues/gh/metamask/metamask-planning/3735)

Question is should we revert the change above and leave this broken until it's fixed properly in `main`? Or should we do a more proper fix here and consider it fixed as part of this PR? 

This PR has too many changes already. I think it makes more sense to undo the "fix" above and leave this broken.

Reply:
done [here](https://github.com/MetaMask/metamask-extension/pull/27847/commits/b08f4e6d6684e824be07a61deed8350b2b020efd)